core/authority-discovery/src/lib.rs

```diff
@@ -0,0 +1,687 @@
+
+		let id = BlockId::hash(self.client.info().best_hash);
+
+		// From the Dht we only get the hashed authority id. In order to retrieve the actual authority id and to ensure
+		// it is actually an authority, we match the hash against the hash of the authority id of all other authorities.
+		let authorities = self.client.runtime_api().authorities(&id)?;
+		self.purge_old_authorities_from_cache(&authorities);
+
+		let authorities = authorities
+			.into_iter()
+			.map(|a| hash_authority_id(a.0.as_ref()).map(|h| (h, a)))
+			.collect::<Result<HashMap<_, _>>>()?;
+
+		for (key, value) in values.iter() {
+			// Check if the event origins from an authority in the current authority set.
+			let authority_id: &AuthorityId = authorities
+				.get(key)
+				.ok_or(Error::MatchingHashedAuthorityIdWithAuthorityId)?;
+
+			let signed_addresses =
+				schema::SignedAuthorityAddresses::decode(value).map_err(Error::Decoding)?;
+
+			let is_verified = self
+				.client
+				.runtime_api()
+				.verify(
+					&id,
+					signed_addresses.addresses.clone(),
+					Signature(signed_addresses.signature.clone()),
+					authority_id.clone(),
+				)
+				.map_err(Error::CallingRuntime)?;
+
+			if !is_verified {
+				return Err(Error::VerifyingDhtPayload);
+			}
+
+			let addresses: Vec<libp2p::Multiaddr> =
+				schema::AuthorityAddresses::decode(signed_addresses.addresses)
+					.map(|a| a.addresses)
+					.map_err(Error::Decoding)?
+					.into_iter()
+					.map(|a| a.parse())
+					.collect::<std::result::Result<_, _>>()
+					.map_err(Error::ParsingMultiaddress)?;
+
+			self.address_cache.insert(authority_id.clone(), addresses);
+		}
+
+		// Let's update the peerset priority group with the all the addresses we have in our cache.
+
+		let addresses = HashSet::from_iter(
+			self.address_cache
+				.iter()
+				.map(|(_peer_id, addresses)| addresses.clone())
+				.flatten(),
+		);
+
+		debug!(target: "sub-authority-discovery", "Applying priority group {:#?} to peerset.", addresses);
+		self.network
+			.set_priority_group("authorities".to_string(), addresses)
+			.map_err(Error::SettingPeersetPriorityGroup)?;
+
+		Ok(())
+	}
+
+	fn purge_old_authorities_from_cache(&mut self, current_authorities: &Vec<AuthorityId>) {
+		self.address_cache
+			.retain(|peer_id, _addresses| current_authorities.contains(peer_id))
+	}
+}
+
+impl<Client, Network, Block> futures::Future for AuthorityDiscovery<Client, Network, Block>
+where
+	Block: BlockT + 'static,
+	Network: NetworkProvider,
+	Client: ProvideRuntimeApi + Send + Sync + 'static + HeaderBackend<Block>,
+	<Client as ProvideRuntimeApi>::Api: AuthorityDiscoveryApi<Block>,
+{
+	type Item = ();
+	type Error = ();
+
+	fn poll(&mut self) -> futures::Poll<Self::Item, Self::Error> {
+		let mut inner = || -> Result<()> {
+			// Process incoming events before triggering new ones.
+			self.handle_dht_events()?;
+
+			if let Ok(Async::Ready(_)) = self.publish_interval.poll() {
```

Comment:
What about timer errors? It seems these are just ignored. I am not sure if the interval would register the task if it encounters an error which means that this `poll` method might not be invoked again, no?

Reply:
According to [Tokio timer documentation](https://github.com/tokio-rs/tokio/blob/master/tokio-timer/src/error.rs) there are two kind of possible errors:

- *Shutdown*: The timer instance is down. Retrying is useless.

- *At capacity*: The timer instance is at its current capacity. It is worth retrying later on.

We could error on the former and implement retry logic with the latter. Retrying would need to be done cautiously e.g. with exponential backoff to not aggravate the timer being overloaded. But how to do exponential backoff without the notion of time?

Without much experience I would expect these errors to be rare.

@twittner: What do you think of logging an error in case one of the two possible timer errors occurs, but not reacting on it? Reducing complexity but still surfacing the fact that there is an issue.